### PR TITLE
feat: post-upgrade tasks

### LINKED ISSUE #838: Support Post Upgrade Task
I've recently been publishing a `tslint-config-<name>` package containing TSLint rules.

Every time I add new rules, or modify rules in a way that requires code changes on the user's part, I release a new major version.

A new major version causes `renovate` to submit pull requests to consuming projects. The pull requests can't be accepted automatically because there are linting errors in the build of consuming projects.

I was wondering if there was a way to setup a process that would run `tslint --fix` on the user's codebase to auto-fix TSLint errors? (This could also apply to `eslint --fix`)

This could also apply to things like [codemods](https://medium.com/@cpojer/effective-javascript-codemods-5a6686bb46fb).

### PR DESCRIPTION
**Pull request by daekene**
Adding support for post-upgrade tasks and for files changed during these tasks to be included in any commits made by Renovate.

The list of tasks are executed as processes in sequence as defined in the config by `postUpgradeTasks`. Once these are complete, the repository is scanned and any files that match the list defined in the `postUpgradeFiles` (a list of Ant matchers) are included in the final commit created by Renovate.

I added this to support running an internal-only Maven plugin which generate dependency lock files. This implementation allows basically any command line tool to be executed by Renovate and any files created or modified by these tools can be included. This obviously opens up some security issues where a repository could put in any command to be executed. Some solutions to avoid this could be:
- Post-upgrade tasks must be selected from a predefined list of allowed tasks
- Post-upgrade tasks can only be executed when the trust level is set to 'high'

I'm open to discussion on how to go ahead with this.

Closes #838

### CODE REVIEW

**Comment by rarkins** [comment]
Hi @daekene, thanks for moving this topic forward. Could you fill out the [CLA](https://cla-assistant.io/renovatebot/renovate?pullRequest=5202) so that I can take a look at the code too?

Security-wise, I think we can roll this feature out first for self-hosted and make it opt-in. That way self-hosted users such as yourself without security concerns can get it right away. Perhaps like this:
 - if trustLevel=high and no limits on tasks, then repositories can define whatever they need
- if trustLevel=high and a set of allowed tasks are configured then it's restricted to that

Defining file paths to match changes against is a good idea. I had been thinking about the same for a future rewrite although as this codebase is Node.js base I'd been leaning towards glob patterns. Would your requirements still be satisfied by that instead of Ant?

**Comment by daekene** [comment]
I've modified this so that in order for a post-upgrade task to be executed, Renovate must be running with `trustLevel` set to high and the command must match one of the regular expressions listed in the `allowedPostUpgradeTasks` list.

I've also swapped out the Ant matchers to use glob patterns instead.

File: docs/usage/self-hosted-configuration.md [hunk 1]
```diff
@@ -7,6 +7,10 @@ description: Self-Hosted Configuration usable in renovate.json or package.json
 
 The below configuration options are applicable only if you are running your own instance ("bot") of Renovate.
 
+## allowedPostUpgradeTasks
```
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md]: Let's add some examples to this description

File: lib/platform/bitbucket-server/bb-got-wrapper.ts [hunk 2]
```diff
@@ -28,7 +28,11 @@ export const api: GotApi = {} as any;
 
 for (const x of helpers) {
   (api as any)[x] = (url: string, opts: any): Promise<GotResponse> =>
-    get(url, { ...opts, method: x.toUpperCase() });
+    get(url, {
+      ...opts,
+      method: x.toUpperCase(),
+      retry: { errorCodes: 'ECONNRESET' },
```
Comment (rarkins) [inline lib/platform/bitbucket-server/bb-got-wrapper.ts]: Can you contribute this as a separate PR?

File: lib/workers/branch/index.ts [hunk 3]
```diff
@@ -320,6 +326,60 @@ export async function processBranch(
     } else {
       logger.debug('No updated lock files in branch');
     }
+
+    if (global.trustLevel === 'high') {
+      logger.debug('Executing post-change hooks');
```
Comment (rarkins) [inline lib/workers/branch/index.ts]:
```suggestion
      logger.debug('Checking for post-upgrade hooks');
```

File: lib/workers/branch/index.ts [hunk 4]
```diff
@@ -320,6 +326,60 @@ export async function processBranch(
...
+            const execResult = await exec(cmd, {
+              cwd: config.localDir,
+            });
+
+            logger.debug(execResult, 'Executed post-change hook');
+          }
+        }
+      }
+
+      if (is.nonEmptyArray(config.postUpgradeFiles)) {
```
Comment (rarkins) [inline lib/workers/branch/index.ts]: I had expected this logic would only be used if a postUpgradeTask was executed. Is it kept separate for any real use case that exists today?

File: lib/workers/branch/index.ts [hunk 5]
```diff
@@ -320,6 +326,60 @@ export async function processBranch(
...
+
+          if (!relativePath.startsWith('.git/')) {
+            for (const pattern of config.postUpgradeFiles) {
+              if (micromatch.isMatch(relativePath, pattern)) {
+                logger.debug(
+                  { file: filePath, pattern },
+                  'Post upgrade file saved'
+                );
+                const existingContent = await readFile(filePath);
+                config.updatedArtifacts.push({
```
Comment (rarkins) [inline lib/workers/branch/index.ts]: Unless I've misunderstood, this will push files regardless of whether they have changed or not. Do you think the logic we have for "vendoring" files (e.g. [in Go Modules](https://github.com/renovatebot/renovate/blob/5b1cd403ab2db660c2c207210efb8a18de218133/lib/manager/gomod/artifacts.ts#L141-L161)) could be used instead? i.e. use git to get a list of changed files then filter them against the list of postUpgradeFiles globs?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md (hunk 1 above)
Comment (daekene) [reply docs/usage/self-hosted-configuration.md]: I have added two examples

File: lib/platform/bitbucket-server/bb-got-wrapper.ts (hunk 2 above)
Comment (daekene) [reply lib/platform/bitbucket-server/bb-got-wrapper.ts]: Yes, I've removed this change from the current PR and I'll raise this later

File: lib/workers/branch/index.ts (hunk 3 above)
Comment (daekene) [reply lib/workers/branch/index.ts]: I've updated the logging statement with your suggestion

File: lib/workers/branch/index.ts (hunk 4 above)
Comment (daekene) [reply lib/workers/branch/index.ts]: Yes, that was the intention. I've shifted this block so it follows the post-upgrade task execution rather than being in a separate block

File: lib/workers/branch/index.ts (hunk 5 above)
Comment (daekene) [reply lib/workers/branch/index.ts]: Not exactly. Renovate will do extra work and store a copy of any matching files in memory, restoring them later. `git add` is then called which would filter out any files which are unchanged. I've followed your suggestion anyway and used the same method as in Go Modules because it will handle deleted files, whereas this does not.

File: docs/usage/configuration-options.md [hunk 6]
```diff
@@ -1080,6 +1080,32 @@ Warning: `pipenv` support is currently in beta, so it is not enabled by default.
...
+Each command must match at least one of the patterns defined in `allowedPostUpgradeTasks` in order to be executed. If the list of allowed tasks is empty then no tasks will be executed.
+
+e.g.
+
+```json
+{
+  "postUpgradeTasks": ["tslint --fix"]
+}
+```
+
```
Comment (ViceIce) [inline docs/usage/configuration-options.md]: Can we join this to a single options object?

File: lib/config/definitions.ts [hunk 7]
```diff
@@ -102,6 +102,31 @@ export type RenovateOptions =
...
 
 const options: RenovateOptions[] = [
+  {
+    name: 'allowedPostUpgradeTasks',
+    description:
+      'A list of regular expressions that determine which post-upgrade tasks are allowed. A task has to match at least one of the patterns to be allowed to run',
+    type: 'array',
+    subType: 'string',
+    default: [],
+    stage: 'global',
```
Comment (ViceIce) [inline lib/config/definitions.ts]:
```suggestion
    admin: true,
```

Should only be changeble by admin.

File: lib/config/index.ts [hunk 8]
```diff
@@ -113,6 +113,8 @@ export async function parseConfigs(
   });
   global.trustLevel = config.trustLevel || 'low';
   delete config.trustLevel;
+  global.allowedPostUpgradeTasks = config.allowedPostUpgradeTasks;
```
Comment (ViceIce) [inline lib/config/index.ts]: I don't like global variables. We should move those to a shared config object or pass that with config as other options.

File: lib/platform/bitbucket-server/bb-got-wrapper.ts [hunk 9]
```diff
@@ -28,7 +28,10 @@ export const api: GotApi = {} as any;
 
 for (const x of helpers) {
   (api as any)[x] = (url: string, opts: any): Promise<GotResponse> =>
-    get(url, { ...opts, method: x.toUpperCase() });
+    get(url, {
+      ...opts,
+      method: x.toUpperCase(),
+    });
```
Comment (ViceIce) [inline lib/platform/bitbucket-server/bb-got-wrapper.ts]: @daekene Can you please remove this unrelated change.

@rarkins was this prettier?

File: lib/types.d.ts [hunk 10]
```diff
@@ -33,6 +33,8 @@ declare namespace NodeJS {
 
     trustLevel?: string;
 
+    allowedPostUpgradeTasks?: string[];
+
```
Comment (ViceIce) [inline lib/types.d.ts]: see above.

File: lib/workers/branch/index.ts [hunk 11]
```diff
@@ -320,6 +325,74 @@ export async function processBranch(
     } else {
       logger.debug('No updated lock files in branch');
     }
+
+    if (global.trustLevel === 'high') {
+      logger.debug('Checking for post-upgrade hooks');
+      if (
+        is.nonEmptyArray(config.postUpgradeTasks) &&
+        is.nonEmptyArray(global.allowedPostUpgradeTasks)
```
Comment (ViceIce) [inline lib/workers/branch/index.ts]:
```suggestion
        is.nonEmptyArray(config.allowedPostUpgradeTasks)
```

If `admin:true` this is not overridable by repo config.

File: lib/workers/branch/index.ts [hunk 12]
```diff
@@ -320,6 +325,74 @@ export async function processBranch(
...
+      ) {
+        for (const cmd of config.postUpgradeTasks) {
+          if (
+            !_.some(global.allowedPostUpgradeTasks, (pattern: string) =>
+              cmd.match(pattern)
+            )
+          ) {
+            logger.debug(
+              { cmd, allowedPostUpgradeTasks: global.allowedPostUpgradeTasks },
```
Comment (ViceIce) [inline lib/workers/branch/index.ts]:
```suggestion
              { cmd, allowedPostUpgradeTasks: config.allowedPostUpgradeTasks },
```

File: lib/workers/branch/index.ts [hunk 13]
```diff
@@ -320,6 +325,74 @@ export async function processBranch(
...
+
+    if (global.trustLevel === 'high') {
+      logger.debug('Checking for post-upgrade hooks');
+      if (
+        is.nonEmptyArray(config.postUpgradeTasks) &&
+        is.nonEmptyArray(global.allowedPostUpgradeTasks)
+      ) {
+        for (const cmd of config.postUpgradeTasks) {
+          if (
+            !_.some(global.allowedPostUpgradeTasks, (pattern: string) =>
```
Comment (ViceIce) [inline lib/workers/branch/index.ts]:
```suggestion
            !_.some(config.allowedPostUpgradeTasks, (pattern: string) =>
```

File: lib/workers/branch/index.ts [hunk 14]
```diff
@@ -320,6 +325,74 @@ export async function processBranch(
...
+        }
+
+        if (is.nonEmptyArray(config.postUpgradeFiles)) {
+          const status = await platform.getRepoStatus();
+
+          for (const filePath of status.modified.concat(status.not_added)) {
+            const relativePath = filePath.substring(config.localDir.length + 1);
+
+            for (const pattern of config.postUpgradeFiles) {
+              if (micromatch.isMatch(relativePath, pattern)) {
```
Comment (ViceIce) [inline lib/workers/branch/index.ts]: @rarkins I think inverting condition and using `continue` would look better?

File: package.json [hunk 15]
```diff
@@ -98,6 +98,7 @@
     "@renovate/pep440": "0.4.1",
     "@sindresorhus/is": "1.2.0",
     "@snyk/ruby-semver": "2.0.4",
+    "@spatocode/dirwalk": "^0.0.1",
```
Comment (ViceIce) [inline package.json]:
```suggestion
```
please remove unsued deps

File: package.json [hunk 16]
```diff
@@ -113,6 +114,7 @@
     "delay": "4.3.0",
     "detect-indent": "6.0.0",
     "email-addresses": "3.1.0",
+    "find": "^0.3.0",
```
Comment (ViceIce) [inline package.json]:
```suggestion
```

File: package.json [hunk 17]
```diff
@@ -155,7 +157,8 @@
     "www-authenticate": "0.6.2",
     "xmldoc": "1.1.2",
     "yarn": "1.21.1",
-    "yawn-yaml": "1.4.0"
+    "yawn-yaml": "1.4.0",
+    "micromatch": "latest"
```
Comment (ViceIce) [inline package.json]:
```suggestion
    "micromatch": "4.0.2"
```
pin version

**Review by viceice — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md (hunk 6 above)
Comment (rarkins) [reply docs/usage/configuration-options.md]: In theory it should not be possible to have files or tasks without the other, right @daekene? If so then ideally the config should be as @ViceIce suggests where the semantics themselves stop someone making the mistake of configuring one and not the other. I'm not sure of exactly how to do that either.

Comment (daekene) [reply docs/usage/configuration-options.md]: @rarkins That's correct, it shouldn't be possible to have tasks without specifying files to include (the intention being that tasks should only be run to include additional files in the commit being created) so the config should tie these two things together.

I've merged these properties into `postUpdateTasks` object, with `commands` and `fileFilters` as two properties of that. 

`allowedPostUpdateCommands` is still separate as that's an administrative setting

File: lib/config/index.ts (hunk 8 above)
Comment (daekene) [reply lib/config/index.ts]: Agreed. When I first read this I thought that 'self-hosted' settings were controlled by this global object. I didn't realise that they could be specified by the regular config object.

File: lib/workers/branch/index.ts (hunk 14 above)
Comment (daekene) [reply lib/workers/branch/index.ts]: That would be my preferred approach as well, but the linter complains about this. Since it's only one condition I've left it as it is but if you still prefer that then I'm happy to change it

File: lib/workers/branch/index.ts [hunk 18]
```diff
@@ -320,6 +326,83 @@ export async function processBranch(
     } else {
       logger.debug('No updated lock files in branch');
     }
+
+    if (global.trustLevel === 'high') {
+      logger.info(
```
Comment (rarkins) [inline lib/workers/branch/index.ts]:
```suggestion
      logger.debug(
```

File: lib/workers/branch/index.ts [hunk 19]
```diff
@@ -320,6 +326,83 @@ export async function processBranch(
     } else {
       logger.debug('No updated lock files in branch');
     }
+
+    if (global.trustLevel === 'high') {
```
Comment (rarkins) [inline lib/workers/branch/index.ts]:
```suggestion
    if (global.trustLevel === 'high' && 
      is.nonEmptyArray(config.allowedPostUpgradeCommands)) {
```

Sorry, it might need prettier-fying too

File: lib/workers/branch/index.ts [hunk 20]
```diff
@@ -320,6 +326,83 @@ export async function processBranch(
...
+          allowedCommands: config.allowedPostUpgradeCommands,
+        },
+        'Checking for post-upgrade tasks'
+      );
+      const commands = config.postUpgradeTasks.commands || [];
+      const fileFilters = config.postUpgradeTasks.fileFilters || [];
+
+      if (
+        is.nonEmptyArray(commands) &&
+        is.nonEmptyArray(config.allowedPostUpgradeCommands)
```
Comment (rarkins) [inline lib/workers/branch/index.ts]: This `allowedPostUpgradeCommands` check can be removed as we check for it earlier

File: lib/workers/branch/index.ts [hunk 21]
```diff
@@ -320,6 +326,83 @@ export async function processBranch(
...
+      ) {
+        for (const cmd of commands) {
+          if (
+            !_.some(config.allowedPostUpgradeCommands, (pattern: string) =>
+              cmd.match(pattern)
+            )
+          ) {
+            logger.debug(
```
Comment (rarkins) [inline lib/workers/branch/index.ts]:
```suggestion
            logger.warn(
```

File: package.json [hunk 22]
```diff
@@ -131,6 +131,7 @@
     "luxon": "1.21.3",
     "markdown-it": "10.0.0",
     "markdown-table": "1.1.3",
+    "micromatch": "4.0.2",
```
Comment (rarkins) [inline package.json]: What caused your preference for micromatch over minimatch?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: package.json (hunk 22 above)
Comment (daekene) [reply package.json]: When searching for glob matchers, the description of micromatch in the NPM registry pitched itself as the more performant alternative to minimatch. I didn't realise that minimatch was already a dependency in Renovate, otherwise I would have just stuck with that. I've reverted back to using minimatch and removed micromatch as a dependency

File: lib/workers/branch/index.ts (hunk 21 above)
Comment (daekene) [reply lib/workers/branch/index.ts]: Changed

File: lib/workers/branch/index.ts (hunk 20 above)
Comment (daekene) [reply lib/workers/branch/index.ts]: Changed

File: lib/workers/branch/index.ts (hunk 18 above)
Comment (daekene) [reply lib/workers/branch/index.ts]: Changed
